samples/boards/esp32/wifi_apsta_mode/src/main.c

```diff
@@ -0,0 +1,190 @@
+			sta_info->mac[2], sta_info->mac[3], sta_info->mac[4], sta_info->mac[5]);
+		break;
+	}
+	default:
+		break;
+	}
+}
+
+static void enable_dhcpv4_server()
+{
+	static struct in_addr addr;
+	static struct in_addr netmaskAddr;
+
+	if (net_addr_pton(AF_INET, CONFIG_WIFI_AP_IP_ADDRESS, &addr)) {
+		LOG_ERR("Invalid address: %s", CONFIG_WIFI_AP_IP_ADDRESS);
+		return;
+	}
+
+	if (net_addr_pton(AF_INET, CONFIG_WIFI_AP_NETMASK, &netmaskAddr)) {
+		LOG_ERR("Invalid netmask: %s", CONFIG_WIFI_AP_NETMASK);
+		return;
+	}
+
+	net_if_ipv4_set_gw(ap_iface, &addr);
+
+	if (net_if_ipv4_addr_add(ap_iface, &addr, NET_ADDR_MANUAL, 0) == NULL) {
+		LOG_ERR("unable to set IP address for AP interface");
+	}
+
+	if (!net_if_ipv4_set_netmask_by_addr(ap_iface, &addr, &netmaskAddr)) {
+		LOG_ERR("Unable to set netmask for AP interface: %s", CONFIG_WIFI_AP_NETMASK);
+	}
+
+	addr.s4_addr[3] += 10; /* Starting IPv4 address for DHCPv4 address pool. */
+
+	if (net_dhcpv4_server_start(ap_iface, &addr) != 0) {
+		LOG_ERR("DHCP server is not started for desired IP");
+		return;
+	}
+
+	LOG_INF("DHCPv4 server started...\n");
+}
+
+static int enable_ap_mode()
+{
+	if (!ap_iface) {
+		LOG_INF("AP: is not initialized");
+		return -EIO;
+	}
+
+	LOG_INF("Turning on AP Mode");
+	ap_config.ssid = (const uint8_t *)CONFIG_WIFI_AP_SSID;
+	ap_config.ssid_length = strlen(CONFIG_WIFI_AP_SSID);
+	ap_config.psk = (const uint8_t *)CONFIG_WIFI_AP_PSK;
+	ap_config.psk_length = strlen(CONFIG_WIFI_AP_PSK);
+	ap_config.channel = WIFI_CHANNEL_ANY;
+	ap_config.band = WIFI_FREQ_BAND_2_4_GHZ;
+
+	if (strlen(CONFIG_WIFI_AP_PSK) == 0) {
+		ap_config.security = WIFI_SECURITY_TYPE_NONE;
+	} else {
+
+		ap_config.security = WIFI_SECURITY_TYPE_PSK;
+	}
+
+	enable_dhcpv4_server();
+
+	int ret = net_mgmt(NET_REQUEST_WIFI_AP_ENABLE, ap_iface, &ap_config,
+			   sizeof(struct wifi_connect_req_params));
+	if (ret) {
+		LOG_ERR("NET_REQUEST_WIFI_AP_ENABLE failed, err: %d", ret);
+	}
+
+	return ret;
+}
+
+static int connect_to_wifi()
+{
+	if (!sta_iface) {
+		LOG_INF("STA: interface no initialized");
+		return -EIO;
+	}
+
+	sta_config.ssid = (const uint8_t *)CONFIG_WIFI_SSID;
+	sta_config.ssid_length = strlen(CONFIG_WIFI_SSID);
+	sta_config.psk = (const uint8_t *)CONFIG_WIFI_PSK;
+	sta_config.psk_length = strlen(CONFIG_WIFI_PSK);
+	sta_config.security = WIFI_SECURITY_TYPE_PSK;
+	sta_config.channel = WIFI_CHANNEL_ANY;
+	sta_config.band = WIFI_FREQ_BAND_2_4_GHZ;
+
+	LOG_INF("Connecting to SSID: %s\n", sta_config.ssid);
+
+	int ret = net_mgmt(NET_REQUEST_WIFI_CONNECT, sta_iface, &sta_config,
+			   sizeof(struct wifi_connect_req_params));
+	if (ret) {
+		LOG_ERR("Unable to Connect to (%s)", CONFIG_WIFI_SSID);
+	}
+
+	return ret;
+}
+
+int main(void)
+{
+	k_sleep(K_SECONDS(5));
+
+	net_mgmt_init_event_callback(&cb, wifi_event_handler, NET_EVENT_WIFI_MASK);
+	net_mgmt_add_event_callback(&cb);
+
+	/* net_if at index 1 is always AP interface in AP-STA mode. */
```

Comment:
Note that this ordering of interfaces is only valid in this sample application. If you enable various other network config, new interfaces might be created and interface ordering could be different.

Reply:
Got it. We will correct the comments. 

Now, I am starting to think that what if more network configuration are enabled. then `net_if` index might get changed. So, how can we make sure that interface we are calling is for `STA` interface and vice versa?  I'm thinking of putting check on node labels to check which interface is which. WDYT think about it? looking for your suggestions.

Reply:
There are these functions in latest upstream. If things are properly configured, they should return the station / AP mode interfaces.
```
/**
 * @brief Get Wi-Fi network station interface.
 *
 * @return Pointer to network interface, NULL if not found.
 */
struct net_if *net_if_get_wifi_sta(void);

/**
 * @brief Get first Wi-Fi network Soft-AP interface.
 *
 * @return Pointer to network interface, NULL if not found.
 */
struct net_if *net_if_get_wifi_sap(void);
```

Reply:
I take a quick look to the reference you provided. Should, I register the interfaces with `wifi_supplicant` WIFI_NM_INSTANCE using `wifi_nm_register_mgd_type_iface` this function from WiFi driver code? So we can `net_if_get_wifi_sap` and `net_if_get_wifi_sta`. Would it be the the right approach?

Reply:
No, NM registration is only needed if you are using a NM (network manager) like WPA suplicant, those APIs handle both NM and non-NM usecases.

Reply:
Both of the function returns the same iface interface. From below code snippet, I think they will return the first iface they found even if there are multiple interfaces when `CONFIG_WIFI_NM` is not set. How can we add the type information without using or defining `NM`?

```
struct net_if *net_if_get_wifi_sta(void)
{
	STRUCT_SECTION_FOREACH(net_if, iface) {
		if (net_if_is_wifi(iface)
#ifdef CONFIG_WIFI_NM
			&& wifi_nm_iface_is_sta(iface)
#endif
			) {
			return iface;
		}
	}

	/* If no STA interface is found, return the first WiFi interface */
	return net_if_get_first_wifi();
}

struct net_if *net_if_get_wifi_sap(void)
{
	STRUCT_SECTION_FOREACH(net_if, iface) {
		if (net_if_is_wifi(iface)
#ifdef CONFIG_WIFI_NM
			&& wifi_nm_iface_is_sap(iface)
#endif
			) {
			return iface;
		}
	}

	/* If no STA interface is found, return the first WiFi interface */
	return net_if_get_first_wifi();
}
`
``

Reply:
Indeed,implementation of both functions would need some enhancements so that they would work with offloaded interfaces where NM is not enabled.

Reply:
@jukkar @krish2718 , can you please, review the code. I have also solved a minor bug related to nm. Now, check if these changes are accepted or not?